commons-hpcc/src/main/java/org/hpccsystems/commons/ecl/RecordDefinitionTranslator.java

```diff
@@ -0,0 +1,733 @@
+                return FieldType.DATASET;
+            default:
+                return FieldType.UNKNOWN;
+        }
+    }
+
+    private static HpccSrcType getSourceType(int typeID)
+    {
+        int type = typeID & TYPE_ID_MASK;
+        switch (type)
+        {
+            case type_int:
+            case type_real:
+                return HpccSrcType.LITTLE_ENDIAN;
+            case type_swapint:
+            case type_biasedswapint:
+            case type_keyedint:
+            case type_filepos:
+                return HpccSrcType.BIG_ENDIAN;
+            case type_utf8:
+                return HpccSrcType.UTF8;
+            case type_char:
+            case type_string:
+            case type_varstring:
+                return HpccSrcType.SINGLE_BYTE_CHAR;
+            case type_qstring:
+                return HpccSrcType.QSTRING;
+            case type_unicode:
+            case type_varunicode:
+                return HpccSrcType.UTF16LE;
+            default:
+                return HpccSrcType.UNKNOWN;
+        }
+    }
+
+    private static boolean isFixedLength(int typeID)
+    {
+        int varLengthBit = typeID & FLAG_UNKNOWN_SIZE;
+        return varLengthBit == 0;
+    }
+
+    private static boolean isUnsigned(int typeID)
+    {
+        return (typeID & FLAG_UNSIGNED) != 0;
+    }
+
+    /**
+     * toECL Converts the provided FieldDef into an ECL record definition
+     *
+     * @param field
+     *            the FieldDef to convert
+     * @return ECL Record defintion as a String
+     * @throws Exception
+     */
+    public static String toECL(FieldDef field) throws Exception
+    {
+        if (field.getFieldType() != FieldType.RECORD)
+        {
+            throw new Exception("Invalid record structure. Root object must of type Record");
+        }
+
+        // Recurse through the tree structure and generate record defintions
+        HashMap<String, String> recordDefinitionMap = new HashMap<String, String>();
+        String rootRecordName = getEClTypeDefinition(field, recordDefinitionMap);
+
+        // Get root record definition and remove it from the map
+        String definition = recordDefinitionMap.get(rootRecordName);
+        recordDefinitionMap.remove(rootRecordName);
+        definition = definition.replace(rootRecordName, "RD");
+
+        // Combine the type definitions into a single ECL defintion
+        int numRecordDefinitions = 1;
+        for (HashMap.Entry<String, String> entry : recordDefinitionMap.entrySet())
+        {
+
+            definition = entry.getKey() + " := " + entry.getValue() + "\n\n" + definition;
+
+            // Replace the temporary hash key with something more readable
+            definition = definition.replace(entry.getKey(), "RD" + numRecordDefinitions);
+            numRecordDefinitions++;
+        }
+
+        return definition;
+    }
+
+    private static String getEClTypeDefinition(FieldDef field, HashMap<String, String> recordDefinitionMap)
```

Comment:
should recordDefinitionMap be fieldDefinitionMap?

Reply:
It should be recordDefinitionMap in this case. It is being used to track the record hash codes -> to record names